cmd/mt-replicator/consume.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/bsm/sarama-cluster"
+	"github.com/raintank/worldping-api/pkg/log"
+	"gopkg.in/raintank/schema.v1"
+)
+
+type Consumer struct {
+	consumer *cluster.Consumer
+	Done     chan struct{}
+}
+
+func NewConsumer(brokers []string, group, topic string) (*Consumer, error) {
+	config := cluster.NewConfig()
+	// see https://github.com/raintank/metrictank/issues/236
+	config.Consumer.Offsets.Initial = sarama.OffsetNewest
+	config.ClientID = "mt-replicator"
+	config.Group.Return.Notifications = true
+	config.ChannelBufferSize = 1000
+	config.Consumer.Fetch.Min = int32(1)
+	config.Consumer.Fetch.Default = int32(32768)
+	config.Consumer.MaxWaitTime = time.Second
+	config.Consumer.MaxProcessingTime = time.Second
+	config.Config.Version = sarama.V0_10_0_0
+	err := config.Validate()
+	if err != nil {
+		return nil, err
+	}
+	consumer, err := cluster.NewConsumer(brokers, group, []string{topic}, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Consumer{
+		consumer: consumer,
+		Done:     make(chan struct{}),
+	}, nil
+}
+
+func (c *Consumer) Consume(publisher *Publisher) {
+	buf := make([]*schema.MetricData, 0)
+	ticker := time.NewTicker(time.Second * 10)
+	counter := 0
+	msgChan := c.consumer.Messages()
+	defer close(c.Done)
+	for {
+		select {
+		case m, ok := <-msgChan:
+			if !ok {
+				return
+			}
+			md := &schema.MetricData{}
+			_, err := md.UnmarshalMsg(m.Value)
+			if err != nil {
+				log.Error(3, "kafka-mdm decode error, skipping message. %s", err)
+				continue
+			}
+			counter++
+			buf = append(buf, md)
+			if len(buf) > 1000 {
+				log.Debug("flushing metricData buffer to kafka.")
+				complete := false
+				for !complete {
+					if err = publisher.Send(buf); err != nil {
```

Comment:
This was a deliberate choice to reduce complexity.  Performance can easily be scaled by just running multiple mt-replicators.